pkg/sql/opt/indexrec/index_recommendation_set.go

```diff
@@ -23,81 +23,141 @@ import (
+// in plan costs is not very significant.
 func FindIndexRecommendationSet(expr opt.Expr, md *opt.Metadata) IndexRecommendationSet {
 	var recommendationSet IndexRecommendationSet
 	recommendationSet.init(md)
 	recommendationSet.createIndexRecommendations(expr)
+	recommendationSet.pruneIndexRecommendations()
 	return recommendationSet
 }
 
 // IndexRecommendationSet stores the hypothetical indexes that are used in a
-// statement's optimal plan (in usedIndexes), as well as the statement's
-// metadata.
+// statement's optimal plan (in indexRecs), as well as the statement's metadata.
 type IndexRecommendationSet struct {
-	md          *opt.Metadata
-	usedIndexes map[cat.Table]util.FastIntSet
+	md        *opt.Metadata
+	indexRecs map[cat.Table][]indexRecommendation
 }
 
 // init initializes an IndexRecommendationSet by allocating memory for it.
 func (irs *IndexRecommendationSet) init(md *opt.Metadata) {
 	numTables := len(md.AllTables())
 	irs.md = md
-	irs.usedIndexes = make(map[cat.Table]util.FastIntSet, numTables)
+	irs.indexRecs = make(map[cat.Table][]indexRecommendation, numTables)
 }
 
 // createIndexRecommendations recursively walks an expression tree to find
 // hypothetical indexes that are used in it.
 func (irs *IndexRecommendationSet) createIndexRecommendations(expr opt.Expr) {
 	switch expr := expr.(type) {
 	case *memo.ScanExpr:
-		irs.addIndexToRecommendationSet(expr.Index, expr.Table)
+		irs.addIndexToRecommendationSet(expr.Index, expr.Cols, expr.Table)
 	case *memo.LookupJoinExpr:
-		irs.addIndexToRecommendationSet(expr.Index, expr.Table)
+		irs.addIndexToRecommendationSet(expr.Index, expr.Cols, expr.Table)
 	case *memo.InvertedJoinExpr:
-		irs.addIndexToRecommendationSet(expr.Index, expr.Table)
+		irs.addIndexToRecommendationSet(expr.Index, expr.Cols, expr.Table)
 	case *memo.ZigzagJoinExpr:
-		irs.addIndexToRecommendationSet(expr.LeftIndex, expr.LeftTable)
-		irs.addIndexToRecommendationSet(expr.RightIndex, expr.RightTable)
+		irs.addIndexToRecommendationSet(expr.LeftIndex, expr.Cols, expr.LeftTable)
+		irs.addIndexToRecommendationSet(expr.RightIndex, expr.Cols, expr.RightTable)
 	}
 	for i, n := 0, expr.ChildCount(); i < n; i++ {
 		irs.createIndexRecommendations(expr.Child(i))
 	}
 }
 
+// pruneIndexRecommendations removes redundant index recommendations from the
+// recommendation set. These are recommendations where the existingIndex field
+// is not nil and no new columns are being stored.
+func (irs *IndexRecommendationSet) pruneIndexRecommendations() {
+	for t, indexRecs := range irs.indexRecs {
+		updatedIndexRecs := make([]indexRecommendation, 0, len(indexRecs))
+		for _, indexRec := range indexRecs {
+			if indexRec.existingIndex == nil || !indexRec.redundantRecommendation() {
+				updatedIndexRecs = append(updatedIndexRecs, indexRec)
+			}
+		}
+		irs.indexRecs[t] = updatedIndexRecs
+	}
+}
+
+// getColOrdSet returns an ordinal set of an opt.ColSet, containing all columns
+// in the set from the table with id tabID.
```

Comment:
nit: This is a bit confusing. Maybe "getColOrdSet returns the set of column ordinals within the given table that are contained in cols."?